File: leetcode/Arrays/leetcode-0059/59.go
package arrays

func generateMatrix(n int) [][]int {
	if n == 0 {
		return [][]int{}
	}
	res := make([][]int, n)
	for i := 0; i < n; i++ {
		res[i] = make([]int, n)
	}

	r, c, row, col := 0, 0, n-1, n-1
	num := 1
	for {
		for i := c; i <= col; i++ {
			res[r][i] = num
			num++
		}
		r++
		if r > row {
			break
		}

		for i := r; i <= row; i++ {
			res[i][col] = num
			num++
		}
		col--
		if c > col {
			break
		}

		for i := col; i >= c; i-- {
			res[row][i] = num
			num++
		}
		row--
		if r > row {
			break
		}

		for i := row; i >= r; i-- {
			res[i][c] = num
			num++
		}
		c++
		if c > col {
			break
		}
	}
	return res
}

File: leetcode/Arrays/leetcode-0059/59.java
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] res = new int[n][n];
        if (n == 0) {
            return res;
        }
        int r = 0, c = 0;
        int row = n - 1, col = n - 1;
        int num = 1;

        while (true) {
            for (int i = c; i <= col; i++) {
                res[r][i] = num++;
            }
            if (++r > row) break;

            for (int i = r; i <= row; i++) {
                res[i][col] = num++;
            }
            if (c > --col) break;

            for (int i = col; i >= c; i--) {
                res[row][i] = num++;
            }
            if (r > --row) break;

            for (int i = row; i >= r; i--) {
                res[i][c] = num++;
            }
            if (++c > col) break;
        }
        return res;
    }
}